sessions-core/src/test/java/com/adyen/checkout/sessions/interactor/SessionInteractorTest.kt

```diff
@@ -0,0 +1,789 @@
+                assertEquals(expectedResult, result)
+
+                assertTrue(sessionInteractor.isFlowTakenOver)
+            }
+
+            @Test
+            fun `merchant doesn't handle call then an exception is thrown and isFlowTakenOver stays true`() = runTest {
+                assertThrows<CheckoutException>(
+                    "Sessions flow was already taken over in a previous call, makeDetails should be implemented"
+                ) {
+                    sessionInteractor.onDetailsCallRequested(ActionComponentData(), { false }, "makeDetails")
+                }
+
+                assertTrue(sessionInteractor.isFlowTakenOver)
+            }
+        }
+    }
+
+    @Nested
+    @DisplayName("when a balance call is requested and")
+    inner class BalanceCallTest {
+
+        @Nested
+        @DisplayName("flow is not taken over and")
+        inner class FlowNotTakenOverTest {
+
+            @Test
+            fun `balance is positive then Successful is returned and session data is updated`() = runTest {
+                sessionInteractor.sessionFlow.test {
+                    val mockResponse = SessionBalanceResponse(
+                        sessionData = "session_data_updated",
+                        balance = Amount("USD", 5000),
+                        transactionLimit = Amount("USD", 1000),
+                    )
+
+                    whenever(sessionRepository.checkBalance(any(), any())) doReturn Result.success(mockResponse)
+
+                    val result = sessionInteractor.checkBalance(TestPaymentMethod(), { false }, "")
+
+                    val expectedResult = SessionCallResult.Balance.Successful(
+                        BalanceResult(
+                            balance = mockResponse.balance,
+                            transactionLimit = mockResponse.transactionLimit,
+                        )
+                    )
+
+                    assertEquals(expectedResult, result)
+
+                    val expectedSessionModel = TEST_SESSION_MODEL.copy(sessionData = mockResponse.sessionData)
+                    assertEquals(expectedSessionModel, expectMostRecentItem())
+                }
+            }
+
+            @Test
+            fun `balance is zero then Error is returned and session data is updated`() = runTest {
+                sessionInteractor.sessionFlow.test {
+                    val mockResponse = SessionBalanceResponse(
+                        sessionData = "session_data_updated",
+                        balance = Amount("USD", 0),
+                        transactionLimit = null,
+                    )
+
+                    whenever(sessionRepository.checkBalance(any(), any())) doReturn Result.success(mockResponse)
+
+                    val result = sessionInteractor.checkBalance(TestPaymentMethod(), { false }, "")
+
+                    assertTrue(result is SessionCallResult.Balance.Error)
+                    require(result is SessionCallResult.Balance.Error)
```

Comment:
Are these `require`s  for smart casting?

Reply:
Yeah, we do them like this elsewhere, do you have a nicer way?

Reply:
Nope, I was just curious